### PR TITLE
[DATA-701] add cartographer integration tests for rdk

### PR DESCRIPTION
**Pull request by JohnN193**
add integration tests for cartographer by grabbing latest binary, as well as adding several other tests for cartographer

A passing integration test is located here https://github.com/viamrobotics/rdk/actions/runs/3706557424

https://viam.atlassian.net/browse/DATA-701

### CODE REVIEW

File: services/slam/builtin/cartographer_int_test.go
Comment (jeremyhyde-viam) [inline services/slam/builtin/cartographer_int_test.go:133]: why doesnt this sleep exist in `orbslam_int_test.go`

File: services/slam/builtin/cartographer_int_test.go
Comment (jeremyhyde-viam) [inline services/slam/builtin/cartographer_int_test.go:64]: [nit] can we reorder this so that is matches orbslam_int_test (i.e. position check then map check)

File: services/slam/builtin/cartographer_int_test.go
Comment (jeremyhyde-viam) [inline services/slam/builtin/cartographer_int_test.go:179]: [nit/opt] You could split off an integration test helper function like in orbslam_int_test for future proofing testing when we are also testing taking in 3d data into cartographer. Or you can wait for that to happen before you split off the helper, up to you

File: services/slam/builtin/cartographer_int_test.go
Comment (jeremyhyde-viam) [inline services/slam/builtin/cartographer_int_test.go:191]: [opt] I thinking the log line is selfexplanatory

File: services/slam/builtin/cartographer_int_test.go
Comment (jeremyhyde-viam) [inline services/slam/builtin/cartographer_int_test.go:257]: see previous comment

File: services/slam/builtin/cartographer_int_test.go
Comment (jeremyhyde-viam) [inline services/slam/builtin/cartographer_int_test.go:305]: [minor] could add a check in here for no new maps being generated

File: services/slam/builtin/cartographer_int_test.go
Comment (jeremyhyde-viam) [inline services/slam/builtin/cartographer_int_test.go:179]: [minor] Add checks for new map file for the test cases where should be one. Also consider checking the timestamp of the map file to see if it aligns with map_rate.

File: services/slam/builtin/cartographer_int_test.go
Comment (jeremyhyde-viam) [inline services/slam/builtin/cartographer_int_test.go:246]: [minor] considering adding in a data_file check to each one that either checks that no new data was added, new data WAS added and/or how many files should be expected

**Review by jeremyhyde-viam — CHANGES_REQUESTED** [review]
Glad to see more int tests for cartographer, I think we can make them a bit beefer by more completely checking the folder state each time with tests for both the map and data folders.

File: services/slam/builtin/cartographer_int_test.go
Comment (JohnN193) [reply services/slam/builtin/cartographer_int_test.go:246]: i do not think this is necessary, since we iterate through a fixed number of pcd scans already in the above `for` loop

File: services/slam/builtin/cartographer_int_test.go
Comment (Otterverse) [reply services/slam/builtin/cartographer_int_test.go:184]: So... if you're actually calling Close(), that close is returning no error, and yet it's still "doing stuff" afterwards, I think that points to a bug, one you should probably hunt down, rather than work around it here.

My guess is that, since there's a lot of background go routines, there's probably an ordering problem between the workgroup sync stuff (activeBackgroundWorkers.) One specific thing that _might_ be it is the ticker in StartDataProcess, which is deferred before the wg.Done() call is deferred, meaning the ticker may still fire AFTER the workgroup is marked done (in a race condition.) I could be wrong on that, but it's possible.

File: services/slam/builtin/cartographer_int_test.go
Comment (JohnN193) [reply services/slam/builtin/cartographer_int_test.go:133]: As far as I am aware, orbslam was not having an issue where maps/files were getting generated during the next test(which is what prompted me to add the sleeps). If that is an issue that is occurring for orbslam then I can add them there as well.

File: services/slam/builtin/cartographer_int_test.go
Comment (tessavitabile) [inline services/slam/builtin/cartographer_int_test.go:26]: [opt] TODO is probably not necessary, since this won't be hard for the implementer of DATA-916 to find.

File: services/slam/builtin/cartographer_int_test.go
Comment (tessavitabile) [inline services/slam/builtin/cartographer_int_test.go:86]: [opt] same here

File: services/slam/builtin/cartographer_int_test.go
Comment (tessavitabile) [inline services/slam/builtin/cartographer_int_test.go:189]: [nit] It's confusing to me that this is called `numMaps` but isn't a number.

File: services/slam/builtin/cartographer_int_test.go
Comment (tessavitabile) [reply services/slam/builtin/cartographer_int_test.go:184]: @Otterverse, I don't think that would be the issue here, since the thing that is "doing stuff" after close is the `carto_grpc_server` binary. This is managed as a `ManagedProcess` and gets shut down by `StopSLAMProcess()`. I think it would be worth taking an hour to investigate what is happening during https://github.com/viamrobotics/goutils/blob/2d1d4397829eed318a3bfb7e2bc8e0427a71c229/pexec/managed_process.go#L287, but if we can't figure it out, I'm okay with this as a workaround. What do you think?

**Review by tessavitabile — COMMENTED** [review]
Looks good!

File: services/slam/builtin/cartographer_int_test.go
Comment (Otterverse) [reply services/slam/builtin/cartographer_int_test.go:184]: I've been working a lot with Managed Process, and it does have some issues stopping (grand) child processes correctly. (I'm working on a PR for that.) That _could_ be a source of it. So if the case that's happening here is for sure that the binary process is what's still creating stuff after Close(), then maybe it is that (or something in the appimage layer(s), ugh.)

That said, if your team is comfortable with a 100ms delay as a workaround, (and the workaround is reliable) I won't block on this. I would like to see it documented as an issue for the future in that case though.

**Review by Otterverse — APPROVED** [review]
LGTM, provided the "binary process is still running after Close()" thing with the sleeps gets a JIRA issue filed.

File: services/slam/builtin/cartographer_int_test.go
Comment (JohnN193) [reply services/slam/builtin/cartographer_int_test.go:184]: yeah looking through the test logs linked above, cartographer never actually prints its "[finishing session](https://github.com/viamrobotics/slam/blob/main/slam-libraries/viam-cartographer/src/main.cc#L14)" log that we expect to see when exiting the server. So I am under the assumption that the Close() call isn't actually working, which is why we get the "[unable to check exit status](https://github.com/viamrobotics/rdk/actions/runs/3687572584/jobs/6241321866#step:6:474)" log. Also worth noting that the timer for the stopping and force exit log does not seem to last 3 seconds like its [configured](https://github.com/viamrobotics/goutils/blob/2d1d4397829eed318a3bfb7e2bc8e0427a71c229/pexec/managed_process.go#L23) to be

Comment (tessavitabile) [reply services/slam/builtin/cartographer_int_test.go:184]: Do you know why we're not getting an error back from Close()?

Comment (Otterverse) [reply services/slam/builtin/cartographer_int_test.go:184]: "unable to check exit status" shows up IF the process got killed instead of closing on its own sometimes. We currently have a hardcoded wait of only 3 seconds between sending SIGINT and then SIGKILL (which leaves no exit code.) However, the kill should happen before the process.Stop() returns. But as noted, it doesn't kill children of children, so it could be orphaning processes.

I'm going to put this in my todo to try to reproduce and look into later this week if I have time.

Comment (Otterverse) [reply services/slam/builtin/cartographer_int_test.go:184]: In the meantime, if you want to try using my "fix" branch in your local testing, you can add this to go.mod

`replace go.viam.com/utils => github.com/Otterverse/goutils b7bf775`

The PR for it is here (not quite finished yet, needs more tests.) https://github.com/viamrobotics/goutils/pull/116

File: services/slam/builtin/cartographer_int_test.go
Comment (kkufieta) [inline services/slam/builtin/cartographer_int_test.go:23]: [nit] `cartoSleepMs`

File: services/slam/builtin/cartographer_int_test.go
Comment (kkufieta) [inline services/slam/builtin/cartographer_int_test.go:133]: [nit] `cartoSleepMs`

File: services/slam/builtin/cartographer_int_test.go
Comment (kkufieta) [reply services/slam/builtin/cartographer_int_test.go:191]: I do actually like these comments. They make it easier to read the code for me.

File: services/slam/builtin/cartographer_int_test.go
Comment (kkufieta) [reply services/slam/builtin/cartographer_int_test.go:257]: I vote for leaving the comment in.

**Review by kkufieta — COMMENTED** [review]
First pass, would like to look at it a second time once all other comments are addressed.

File: services/slam/builtin/cartographer_int_test.go
Comment (JohnN193) [reply services/slam/builtin/cartographer_int_test.go:184]: the go.mod file is telling me I need to have the version in the form of `v1.2.3` instead of  `b7bf775`, what should I put for that?

File: services/slam/builtin/cartographer_int_test.go
Comment (JohnN193) [reply services/slam/builtin/cartographer_int_test.go:179]: added a check for the offline case.

File: services/slam/builtin/cartographer_int_test.go
Comment (JohnN193) [reply services/slam/builtin/cartographer_int_test.go:191]: had the comment because there was one initially for the offline mode test

File: services/slam/builtin/cartographer_int_test.go
Comment (Otterverse) [reply services/slam/builtin/cartographer_int_test.go]: Doh! Forgot when you use that short syntax, you have to run `go mod tidy` and then it'll download and correct the syntax. (I'm just in the habit of always running that after I modify go.mod)

File: go.mod
Comment (JohnN193) [inline go.mod:374]: will remove before merging

File: go.sum
Comment (JohnN193) [inline go.sum:109]: will remove before merging

**Review by tessavitabile — APPROVED** [review]

**Review by jeremyhyde-viam — APPROVED** [review]
LGTM, added one more thought to a discussion regarding checking the map state but am open to implementing it later or being convinced it is not necessary for this type of testing.

File: services/slam/builtin/cartographer_int_test.go
Comment (kkufieta) [inline services/slam/builtin/cartographer_int_test.go:130]: added -> Added

File: services/slam/builtin/cartographer_int_test.go
Comment (kkufieta) [inline services/slam/builtin/cartographer_int_test.go:181]: added -> Added

File: services/slam/builtin/cartographer_int_test.go
Comment (kkufieta) [inline services/slam/builtin/cartographer_int_test.go:252]: added -> Added

**Review by kkufieta — APPROVED** [review]
LGTM mod perfectionistic nits

File: services/slam/builtin/cartographer_int_test.go
Comment (JohnN193) [reply services/slam/builtin/cartographer_int_test.go:303]: spoke with jeremy on this, and added a function to test the data_dir after tests

**Review by jeremyhyde-viam — APPROVED** [review]
LGTM

**Pull request merged**
